### ISSUE TITLE
Erroneous refactoring

### ISSUE THREAD

**Issue by Aster89** [haskell/haskell-language-server#2791]
### Your environment

Which OS do you use: up-to-date ArchLinux
Which LSP client (editor/plugin) do you use: terminal Vim + YCM + GHC 8.10.7 + HLS 1.6.1.0
Describe your project (alternative: link to the project): no project, I found the bug during a coding challenge

### Steps to reproduce
Write
```haskell
f = const [] . (>>= const Nothing) . const Nothing
```
and apply the automatic refactoring

### Expected behaviour
The original code should become
```haskell
f = const [] . (const Nothing <=< const Nothing)
```

### Actual behaviour
The original code becomes
```haskell
f = const [] . const Nothing <=< const Nothing
```
which is type-wise erroneous.

### Include debug information

[It is not a YCM problem.](https://github.com/ycm-core/YouCompleteMe/issues/4020#issuecomment-1073269173)

**Comment by michaelpj** [comment]
Does this occur if you use `hlint --refactor` at the command line? If so, it's a hlint bug.

**Comment by Aster89** [comment]
Yep, it gets it wrong. However, since `hlint code.hs` gives the correct suggestion, while `hlint --refactor code.hs` applies it the wrong way, and since I could not run the latter before `cabal install apply-refact`, I suppose I should file the bug report [here](https://github.com/mpickering/apply-refact/issues).

**Comment by michaelpj** [comment]
Well, either that or hlint should be giving a hint that has parentheses. I don't know where the responsibility lies between the two.
